ibis/expr/operations/arrays.py

```diff
@@ -7,23 +7,32 @@
 import ibis.expr.datashape as ds
 import ibis.expr.datatypes as dt
 import ibis.expr.rules as rlz
-from ibis.common.annotations import attribute
+from ibis.common.annotations import ValidationError, attribute
 from ibis.common.typing import VarTuple  # noqa: TCH001
 from ibis.expr.operations.core import Unary, Value
 
 
 @public
 class Array(Value):
     exprs: VarTuple[Value]
+    dtype: Optional[dt.Array] = None
```

Comment:
I think I would prefer to split this operation into two to avoid this messy validation.

One op for inferred type and one for specified type. As is this logic seems unnecessarily complicated.

Reply:
If the complex validation stuff gets hoisted into ibis.array() (see below), do you still have this concern?